pkg/server/events.go

```diff
@@ -175,3 +143,89 @@ func (em *eventMonitor) handleEvent(evt *events.Envelope) {
 		}
 	}
 }
+
+// handleContainerExit handles TaskExit event for container.
+func handleContainerExit(e *eventtypes.TaskExit, cntr containerstore.Container) {
+	if e.Pid != cntr.Status.Get().Pid {
+		// Non-init process died, ignore the event.
+		return
+	}
+	// Attach container IO so that `Delete` could cleanup the stream properly.
+	task, err := cntr.Container.Task(context.Background(),
+		func(*containerdio.FIFOSet) (containerdio.IO, error) {
+			return cntr.IO, nil
+		},
+	)
+	if err != nil {
+		if !errdefs.IsNotFound(err) {
+			logrus.WithError(err).Errorf("failed to load task for container %q", e.ContainerID)
+			return
+		}
+	} else {
+		// TODO(random-liu): [P1] This may block the loop, we may want to spawn a worker
+		if _, err = task.Delete(context.Background()); err != nil {
+			// TODO(random-liu): [P0] Enqueue the event and retry.
+			if !errdefs.IsNotFound(err) {
+				logrus.WithError(err).Errorf("failed to stop container %q", e.ContainerID)
+				return
+			}
+			// Move on to make sure container status is updated.
+		}
+	}
+	err = cntr.Status.UpdateSync(func(status containerstore.Status) (containerstore.Status, error) {
+		// If FinishedAt has been set (e.g. with start failure), keep as
+		// it is.
+		if status.FinishedAt != 0 {
+			return status, nil
+		}
+		status.Pid = 0
+		status.FinishedAt = e.ExitedAt.UnixNano()
+		status.ExitCode = int32(e.ExitStatus)
+		return status, nil
+	})
+	if err != nil {
+		logrus.WithError(err).Errorf("Failed to update container %q state", e.ContainerID)
+		// TODO(random-liu): [P0] Enqueue the event and retry.
+		return
+	}
+}
+
+// handleSandboxExit handles TaskExit event for sandbox.
+func handleSandboxExit(e *eventtypes.TaskExit, sb sandboxstore.Sandbox) {
+	// Don't need to check pid because sandbox container should only have one process.
+	// No stream attached to sandbox container.
+	task, err := sb.Container.Task(context.Background(), nil)
```

Comment:
I  think we should not do this  on unknown state.  Let RunPodSandbox do it. 
Or else we maybe do task.Delete both in handleSandboxExit and RunPodSandbox.

Reply:
On a second thought, I still prefer put `task.Delete` out of `Update.

`Update` will hold a lock of the container/sandbox, and `List` (which is a critical operation for kubelet) will be blocked until the lock is released.

We should put as little logic as possible into `Update`. The issue you mentioned can be handled by https://github.com/containerd/cri-containerd/pull/571#discussion_r163756316.